frontend/src/common/backend.ts

```diff
@@ -163,6 +163,52 @@ export type VersionDto = {
   keycloakVersion: string;
 }
 
+enum AuditEventType {
+  Unlock = 'UNLOCK'
+}
+
+export abstract class AuditEventDto {
+  constructor(public id: number, public timestamp: Date, public type: AuditEventType) { }
+
+  abstract details(): { [key: string]: string };
+}
+
+enum UnlockResult {
+  Success = 'SUCCESS',
+  NoSuchDevice = 'NO_SUCH_DEVICE',
+  DeviceNotAuthorized = 'DEVICE_NOT_AUTHORIZED'
+}
+
+export class UnlockEventDto extends AuditEventDto {
+  constructor(public id: number, public timestamp: Date, public type: AuditEventType, public userId: string, public vaultId: string, public deviceId: string, public result: UnlockResult) {
+    super(id, timestamp, type);
+  }
+
+  details(): { [key: string]: string } {
+    return {
+      userId: this.userId,
+      vaultId: this.vaultId,
+      deviceId: this.deviceId,
+      result: this.result
+    };
+  }
+
+  static typeOf(obj: any): obj is UnlockEventDto {
+    const unlockEventDto = obj as UnlockEventDto;
+    return typeof unlockEventDto.id === 'number'
+      && typeof unlockEventDto.timestamp === 'string'
+      && typeof unlockEventDto.userId === 'string'
+      && typeof unlockEventDto.vaultId === 'string'
+      && typeof unlockEventDto.deviceId === 'string'
+      && Object.values(UnlockResult).includes(unlockEventDto.result)
+      && unlockEventDto.type === AuditEventType.Unlock;
+  }
+
+  static copy(obj: UnlockEventDto): UnlockEventDto {
+    return new UnlockEventDto(obj.id, new Date(obj.timestamp), obj.type, obj.userId, obj.vaultId, obj.deviceId, obj.result);
+  }
+}
```

Comment:
Your suggestion didn't quite work because `details` is not part of the DTO. It was a method I added that only existed in the frontend code. In the backend, `UnlockEventDto` is an extension of `AuditEventDto`.

My goal was to show only the key/values of the subtype in the details column.

I did the refactoring in affd91798e270961e87804fb03a3e88cdc4f5613. Does it make sense?

- `AuditEventDto` is now a simple type and `UnlockEventDto` is an extension of `AuditEventDto` (by extending the type via intersections).
- `AuditLogService.getAllEvents()` doesn't map anything specific anymore.
- `AuditLog` component now has some code to compute the `auditEventDetails`, which transforms the audit events by filtering out the `AuditEventDto` properties.

An alternative would be to somehow change the backend to return a DTO that matches your suggestion. But I didn't know if that would make sense and if it would complicate things in the backend. I'm open for suggestions.

Reply:
current impl looks very clean